bugsnag-plugin-android-ndk/src/main/jni/bugsnag.c

```diff
@@ -13,58 +13,125 @@
 #include <stdlib.h>
 #include <string.h>
 
-static JNIEnv *bsg_global_jni_env = NULL;
+// Acquire a java environment.
+// You must call release_env() when finished, passing in requires_detach.
+// requires_detach will always be correctly set, even on error.
+static JNIEnv *acquire_env(const char *purpose, bool *requires_detach) {
+  *requires_detach = false;
+  if (!bsg_jni_cache->initialized) {
+    BUGSNAG_LOG("Failed to %s: JNI cache has not been initialized via "
+                "bugsnag_start() or NativeBridge.install()",
+                purpose);
+    return NULL;
+  }
+
+  JNIEnv *env = NULL;
+  jint state = (*bsg_jni_cache->jvm)
+                   ->GetEnv(bsg_jni_cache->jvm, (void **)&env, JNI_VERSION_1_6);
+  switch (state) {
+  case JNI_EDETACHED:
+    if ((*bsg_jni_cache->jvm)
+            ->AttachCurrentThread(bsg_jni_cache->jvm, &env, NULL) != JNI_OK) {
+      BUGSNAG_LOG("Failed to %s: Could not attach to JVM thread", purpose);
+      return NULL;
+    }
```

Comment:
Is it worth using a pthread destructor to ensure we detach the thread on termination?
https://github.com/bugsnag/bugsnag-plugin-android-cxxapi/blob/master/src/main/cpp/utils/JavaLoader.cpp#L34

Reply:
I think that code is to handle the normal case where a thread you're calling from will eventually terminate?
This particular thread will never terminate until forcibly shut down on app termination, so there'd be no JVM to detach at that point, which might cause a crash (I think? These are pretty strange waters).

Reply:
Its for either case. this isn't _that_ weird. it makes sense to tidy up and detach in the case that the thread somehow terminates so it doesn't take the whole app with it, even for the case that we revisit this later to adjust how/when the thread is started and ends.

Reply:
My worry is the possibility that JVM cleanup happens before thread reaping when the app terminates. In that case, there could be a danger of the env pointer or related JVM pointers being invalid by the time the thread cleanup code runs.

If we don't make it clean up the env, it would just leak memory, wouldn't it?

Reply:
The strategy in question is used in unreal engine and a few other places, its not dangerous.

Reply:
Failing to detach a thread that terminates will cause an app-level crash.

Reply:
Hang on... I'm thinking of the wrong PR. This code is part of a two-function api. Anyone calling `acquire_env` must call `release_env`, which handles detaching. It's used by the non-env functions `bugsnag_notify`, `bugsnag_set_user`, and `bugsnag_leave_breadcrumb`, which all call the release function.

Reply:
These would still cause problems if the thread was externally terminated for any reason (thinking `pthread_cancel` or similar).